core/Pos/Core/Genesis/Types.hs

```diff
@@ -135,16 +142,58 @@ mkGenesisDelegation psks = do
         throwError "one of the delegates is also an issuer, don't do it"
     return $ UnsafeGenesisDelegation resMap
 
+data TestBalanceOptions = TestBalanceOptions
+    { tsoPoors        :: Word
+    , tsoRichmen      :: Word
+    , tsoRichmenShare :: Double
+    , tsoTotalBalance :: Word64
+    } deriving (Show)
+
+data FakeAvvmOptions = FakeAvvmOptions
+    { faoCount      :: Word
+    , faoOneBalance :: Word64
+    } deriving (Show)
+
+data TestnetDistribution
+    = TestnetBalanceStakeDistr
+    | TestnetRichmenStakeDistr
+    | TestnetCustomStakeDistr
+    { tcsdBootStakeholders :: !GenesisWStakeholders
+    , tcsdVssCerts         :: !VssCertificatesMap
+    } deriving (Show)
+
+data GenesisInitializer
+    = TestnetInitializer {
+      tiTestBalance     :: !TestBalanceOptions
+    , tiFakeAvvmBalance :: !FakeAvvmOptions
+    -- ^ Explicit bootstrap era stakeholders, list of addresses with
+    -- weights (@[(A, 5), (B, 2), (C, 3)]@). Setting this
+    -- overrides default settings for boot stakeholders (e.g. rich
+    -- in testnet stakes).
+    , tiDistribution    :: !TestnetDistribution
+    , tiSeed            :: !Integer
+      -- ^ Seed to use (when no seed is provided, a secure random generator
+      -- is used)
+    }
+    | MainnetInitializer {
+      miStartTime        :: !Timestamp
```

Comment:
No start time needed for the testnet initializer?

Reply:
Right, to be passed in runtime

Reply:
Conceptually the testnet initializer still needs a start time so we should put it in that constructor as well. What will we do with the command-line option for start time in dev mode? Use it to construct the initializer I'd imagine.